File: clients/examples/modern/java/Subscribe.java
// Subscribe.java — receive data from the goat fabric (modern Java, JDK 17+).
//
//   ./gradlew run -Pmain=Subscribe                                   # your namespace, forever
//   ./gradlew run -Pmain=Subscribe --args="release/goat/**"   # inbound data from goat
//   ./gradlew run -Pmain=Subscribe --args="<keyexpr> 5"              # exit after 5 samples
//
// Default key-expr is '<namespace>/**' (everything under your prefix). Use ** for any depth,
// * for a single segment.

import java.nio.charset.StandardCharsets;
import java.time.LocalTime;
import java.time.format.DateTimeFormatter;
import java.util.concurrent.atomic.AtomicInteger;

import io.zenoh.Session;
import io.zenoh.bytes.ZBytes;
import io.zenoh.keyexpr.KeyExpr;
import io.zenoh.pubsub.Subscriber;
import io.zenoh.sample.Sample;

public final class Subscribe {

    private static final DateTimeFormatter TIME = DateTimeFormatter.ofPattern("HH:mm:ss");

    public static void main(String[] args) throws Exception {
        String keyexpr = args.length > 0 ? args[0] : EfdiConnect.namespace() + "/**";
        int limit = args.length > 1 ? Integer.parseInt(args[1]) : 0; // 0 = follow forever
        AtomicInteger seen = new AtomicInteger(0);

        try (Session session = EfdiConnect.session();
             KeyExpr keyExpr = KeyExpr.tryFrom(keyexpr)) {
            Subscriber subscriber = session.declareSubscriber(keyExpr, sample -> onSample(sample, seen));
            System.out.println("subscribed: " + keyexpr + " (Ctrl-C to stop)");
            System.out.flush();
            while (limit == 0 || seen.get() < limit) {
                Thread.sleep(200);
            }
        }
    }

    private static void onSample(Sample sample, AtomicInteger seen) {
        ZBytes payload = sample.getPayload();
        byte[] raw = payload.toBytes();
        String shown;
        if (isUtf8(raw)) {
            shown = new String(raw, StandardCharsets.UTF_8);
        } else {
            shown = "<" + raw.length + " bytes> " + hexPrefix(raw, 32);
        }
        System.out.println(LocalTime.now().format(TIME) + "  " + sample.getKeyExpr() + "  " + shown);
        System.out.flush();
        seen.incrementAndGet();
    }

    /** Strict UTF-8 check: decode and compare round-trip length, falling back to hex on garbage. */
    private static boolean isUtf8(byte[] b) {
        try {
            StandardCharsets.UTF_8.newDecoder()
                .decode(java.nio.ByteBuffer.wrap(b));
            return true;
        } catch (java.nio.charset.CharacterCodingException e) {
            return false;
        }
    }

    private static String hexPrefix(byte[] b, int max) {
        int len = Math.min(b.length, max);
        StringBuilder sb = new StringBuilder(len * 2);
        for (int i = 0; i < len; i++) {
            sb.append(String.format("%02x", b[i]));
        }
        return sb.toString();
    }
}

File: clients/examples/modern/rust/src/bin/subscribe.rs
//! subscribe.rs — receive data from the EFDI fabric (modern Rust).
//!
//! ```text
//! cargo run --bin subscribe                              # your own namespace, follow forever
//! cargo run --bin subscribe -- 'release/<partner>/**'  # inbound data a partner sends you
//! cargo run --bin subscribe -- '<keyexpr>' 5             # exit after 5 samples
//! ```
//!
//! Default key-expr is `<namespace>/**` (everything under your prefix). Use `**` for any
//! depth, `*` for a single segment.

#[tokio::main]
async fn main() -> efdi_connect::Result<()> {
    zenoh::init_log_from_env_or("error");

    let args: Vec<String> = std::env::args().collect();
    let keyexpr = match args.get(1) {
        Some(k) => k.clone(),
        None => format!("{}/**", efdi_connect::namespace()?),
    };
    let limit: usize = args.get(2).and_then(|s| s.parse().ok()).unwrap_or(0); // 0 = forever

    let session = efdi_connect::session().await?;
    let subscriber = session.declare_subscriber(keyexpr.clone()).await?;
    println!("subscribed: {keyexpr} (Ctrl-C to stop)");

    let mut seen: usize = 0;
    loop {
        tokio::select! {
            recv = subscriber.recv_async() => {
                let sample = match recv {
                    Ok(s) => s,
                    Err(_) => break, // subscriber closed
                };
                let raw = sample.payload().to_bytes();
                let shown = match sample.payload().try_to_string() {
                    Ok(s) => s.into_owned(),
                    Err(_) => {
                        let head: Vec<u8> = raw.iter().take(32).copied().collect();
                        format!("<{} bytes> {}", raw.len(), hex(&head))
                    }
                };
                println!(
                    "{}  {}  {}",
                    chrono::Local::now().format("%H:%M:%S"),
                    sample.key_expr().as_str(),
                    shown
                );
                seen += 1;
                if limit != 0 && seen >= limit {
                    break;
                }
            }
            _ = tokio::signal::ctrl_c() => break,
        }
    }

    Ok(())
}

fn hex(bytes: &[u8]) -> String {
    bytes.iter().map(|b| format!("{b:02x}")).collect()
}
